### PR TITLE
ABC flow (synthesis, mapping) supported

### LINKED ISSUE #17: Building & testing ABC
As decided we will rely on `ABC` for logic synthesis and technology mapping, we need to do the following:
- [x] Provide instructions in `README` for building `ABC` on different platforms.
- [x] Build `ABC` on Travis and AppVeyor using these instructions.
- [x] Include testing `ABC` functionality into automated testsuite.

### PR DESCRIPTION
**Pull request by allegroCoder**
- Building & testing ABC on Travis and AppVeyor (fixed #17 )
- `Espresso` related flow for synthesis of the Boolean equations removed. It is replaced with `ABC`.
- `README` file updated with instructions for installing `ABC`, and for running `Scenco` under all the OSes.
- Added support for area estimation, verilog file generation.
- All the test present in https://github.com/allegroCoder/graph-families added and tested.
- `Formulae` as an entity composed by a combination of three lists {input, output, equations}.
- Removed log file generated by `SCENCO`
- Temporary files remove safely in case of any undefined errors.

### CODE REVIEW

File: src/Tuura/Formula.hs [hunk 1]
```diff
@@ -1,3 +1,26 @@
...
+foreign import ccall unsafe "free_formulae"
+    freeFormulae :: IO ()
+
+-- To use once Formula entity will be defined
+--generateFormulae :: FilePath -> IO [Formula]
+--generateFormulae espressoPath = undefined
+
+generateController :: FilePath-> IO Int
+generateController espressoPath = do
+    espresso <- newCString espressoPath
```
Comment (snowleopard) [inline src/Tuura/Formula.hs]: Why are we still using `espresso`? I thought with `ABC` we will no longer need it.

Comment (allegroCoder) [inline src/Tuura/Formula.hs]: Just forgot to modify the names. :+1: I commited a new change, thanks

File: Test.hs [hunk 2]
```diff
@@ -64,53 +88,56 @@ testEncoding (x:xs) (y:ys) = do
...
+    area <- estimateArea abcPath techLibPath
+    putStrLn ("\tArea of the controller: " ++ show area)
+--  resultV <- writeVerilog abcPath techLibPath verilogPath
+--  check resultV "\tVerilog file generation: OK\n" "Verilog file generation: ERROR"
+    unloadController
+
 check :: Int -> String -> String -> IO ()
 check result msgOk msgError
-    | result == 0 = putStrLn msgOk
+    | result == 0 = putStr msgOk
```
Comment (snowleopard) [inline Test.hs]: Why this change? As I see you now had to append `\n` on all call sites?

File: Test.hs [hunk 3]
```diff
@@ -64,53 +88,56 @@ testEncoding (x:xs) (y:ys) = do
...
 assertUnload = do
     result <- unloadGraphsAndCodes
-    check result "Graphs and codes unloaded" "Error unloading graphs"
+    check result "Graphs and codes unloaded\n" "Error unloading graphs"
 
 assertSingleLiteral :: IO ()
 assertSingleLiteral = do
     result <- encodeGraphs SingleLiteral Nothing
-    check result "Single literal encoding: OK" "Single literal encoding: ERROR"
+    check result "> Single literal encoding: OK\n" "Single literal encoding: ERROR"
```
Comment (snowleopard) [inline Test.hs]: Why is this case different from above? Does extra `>` have any meaning? I'd prefer to keep all cases consistent.

File: src/Tuura/Formula.hs [hunk 4]
```diff
@@ -1,3 +1,26 @@
...
+
+generateController :: FilePath-> IO Int
+generateController abcPath = do
+    abc <- newCString abcPath
+    getFormulae abc 1
+
+generateCPOG :: FilePath -> IO Int
+generateCPOG abcPath = do
+    abc <- newCString abcPath
+    getFormulae abc 0
```
Comment (snowleopard) [inline src/Tuura/Formula.hs]: This file needs a lot of further refactoring (getting rid of magic integers, etc.). But we can keep it as is for now.

File: src/Tuura/Synthesis.hs [hunk 5]
```diff
@@ -1,7 +1,17 @@
-module Tuura.Synthesis (synthesis) where
+module Tuura.Synthesis (synthesis, SynthesisType (..),
+                        unloadController) where
 
 import Tuura.Formula
-import Tuura.Graph
 
-synthesis :: Graph -> [Formula]
```
Comment (snowleopard) [inline src/Tuura/Synthesis.hs]: Why are you changing the API we agreed upon previously? `synthesis` should be a function which takes a `Graph` and produces a list of Boolean formulas. Instead the result of synthesis is not an integer! This makes no sense.

File: src/Tuura/TechnologyMapping.hs [hunk 6]
```diff
@@ -1,13 +1,32 @@
...
 technologyMapping = undefined
 
-data Area = Area Double
+estimateArea :: FilePath -> FilePath -> IO (Double)
+estimateArea abcPath techLibPath = do
+    abc <- newCString abcPath
+    techLib <- newCString techLibPath
+    mapAndGetAreaCircuit abc techLib
 
-estimateArea :: Library -> [Formula] -> Area
```
Comment (snowleopard) [inline src/Tuura/TechnologyMapping.hs]: Again, why are you changing the API? Just compare how clear was this type:
```haskell
estimateArea :: Library -> [Formula] -> Area
```
compared to what you are suggesting:
```haskell
estimateArea :: FilePath -> FilePath -> IO (Double)
```
What are these two files? Why files at all? We should have datatypes for representing a `Library` and a list of Boolean formulas. Maybe a Library could be defined just as
```haskell
data Library = Library FilePath
```
But this doesn't mean that we want to use `FilePath`'s everywhere. We might want to extend the above definition of `Library` later without breaking things, but we will never be able to extend the definition of `FilePath`.

File: src/Tuura/Synthesis.hs (hunk 5 above)
Comment (allegroCoder) [inline src/Tuura/Synthesis.hs]: You are right, but we don't have neither the `Graph` entity, nor the `Formula` so far. I created those functions for testing purpose. The integer is just for understanding wether the function runs correctly. Once that all those entities are done, we could use them. Let's discuss about this at lunch?

**Comment by snowleopard** [comment]
I added a bunch of comments above :)

The main issue is that for some reason you are changing a perfectly clear strongly typed API with some error-prone and obscure one, where parameters are stored in files? This is even worse than having `String` parameters everywhere. Please bring the nice API we had back!

File: src/Tuura/Synthesis.hs (hunk 5 above)
Comment (snowleopard) [inline src/Tuura/Synthesis.hs]: The simple way around this may be as follows (similarly to what I suggested about `Library`). Create a wrapper datatype:
```haskell
data Graph = Graph FilePath
```
In this way you hide the current implementation (that a graph is actually somehow represented by a file) and provide an ability to change this later. Don't export the constructor of the datatype, so others will not be able to rely on the fact that it is actually just a file.

File: Test.hs (hunk 2 above)
Comment (allegroCoder) [inline Test.hs]: Not over all the cases, that is the main reason why I changed this. However, it was just for having `valid encoding` on the same line of `... encoding`

File: Test.hs (hunk 3 above)
Comment (allegroCoder) [inline Test.hs]: Ok, I will change this

**Comment by allegroCoder** [comment]
@snowleopard now you can have a look at the PR for merging. Please, let me know if you want me to modify anything else.
